src/assembler.jl

```diff
@@ -89,71 +100,92 @@ Assembles the element residual `ge` into the global residual vector `g`.
     end
 end
 
-abstract type AbstractSparseAssembler end
-
 """
     matrix_handle(a::AbstractSparseAssembler)
     vector_handle(a::AbstractSparseAssembler)
 
-Return a reference to the underlying matrix/vector of the assembler.
+Return a reference to the underlying matrix/vector of the assembler used during
+assembly operations.
 """
 matrix_handle, vector_handle
 
-struct AssemblerSparsityPattern{Tv,Ti} <: AbstractSparseAssembler
-    K::SparseMatrixCSC{Tv,Ti}
+"""
+    get_matrix(a::AbstractSparseAssembler)
+    get_vector(a::AbstractSparseAssembler)
+
+Return a reference to the matrix into which we assemble.
+The returned object might differ from [`matrix_handle`](@ref).
+"""
+get_matrix, get_vector
+
+get_vector(a::AbstractSparseAssembler) = a.f
+get_matrix(a::AbstractSparseAssembler) = a.K
+
+"""
+Assembler for sparse matrix with CSC storage type.
+"""
+struct CSCAssembler{Tv,Ti,MT<:AbstractSparseMatrixCSC{Tv,Ti}} <: AbstractSparseCSCAssembler
+    K::MT
     f::Vector{Tv}
     permutation::Vector{Int}
     sorteddofs::Vector{Int}
 end
-struct AssemblerSymmetricSparsityPattern{Tv,Ti} <: AbstractSparseAssembler
-    K::Symmetric{Tv,SparseMatrixCSC{Tv,Ti}}
+
+"""
+Assembler for symmetric sparse matrix with CSC storage type.
+"""
+struct SymmetricCSCAssembler{Tv,Ti, MT <: Symmetric{Tv,<:AbstractSparseMatrixCSC{Tv,Ti}}} <: AbstractSparseCSCAssembler
+    K::MT
     f::Vector{Tv}
     permutation::Vector{Int}
     sorteddofs::Vector{Int}
 end
 
-function Base.show(io::IO, ::MIME"text/plain", a::Union{AssemblerSparsityPattern,AssemblerSymmetricSparsityPattern})
+function Base.show(io::IO, ::MIME"text/plain", a::AbstractSparseAssembler)
     print(io, typeof(a), " for assembling into:\n - ")
-    summary(io, a.K)
-    if !isempty(a.f)
+    summary(io, get_matrix(a))
+    f = get_vector(a)
+    if !isempty(f)
         print(io, "\n - ")
-        summary(io, a.f)
+        summary(io, f)
     end
 end
 
-matrix_handle(a::AssemblerSparsityPattern) = a.K
-matrix_handle(a::AssemblerSymmetricSparsityPattern) = a.K.data
-vector_handle(a::Union{AssemblerSparsityPattern, AssemblerSymmetricSparsityPattern}) = a.f
+matrix_handle(a::AbstractSparseCSCAssembler) = a.K
+matrix_handle(a::SymmetricCSCAssembler) = a.K.data
+vector_handle(a::AbstractSparseCSCAssembler) = a.f
 
 """
-    start_assemble(K::SparseMatrixCSC;            fillzero::Bool=true) -> AssemblerSparsityPattern
-    start_assemble(K::SparseMatrixCSC, f::Vector; fillzero::Bool=true) -> AssemblerSparsityPattern
+    start_assemble(K::AbstractSparseMatrixCSC;            fillzero::Bool=true) -> CSCAssembler
+    start_assemble(K::AbstractSparseMatrixCSC, f::Vector; fillzero::Bool=true) -> CSCAssembler
```

Comment:
```suggestion
    start_assemble(K::AbstractSparseMatrixCSC;            fillzero::Bool=true, sizehint::Int=0) -> CSCAssembler
    start_assemble(K::AbstractSparseMatrixCSC, f::Vector; fillzero::Bool=true, sizehint::Int=0) -> CSCAssembler
```
Or do you intend to keep this as private API for now? (In that case, perhaps add a note about this?) If public, perhaps `workspace_sizehint` is better even if longer, not to confuse with the sizehint supplied to `start_assemble(N)`?